python/tank/util/shotgun/publish_resolve.py

```diff
@@ -0,0 +1,469 @@
+
+    # url = "file:///path/to/some/file.txt"
+    # ParseResult(
+    #   scheme='file',
+    #   netloc='',
+    #   path='/path/to/some/file.txt',
+    #   params='',
+    #   query='',
+    #   fragment=''
+    # )
+
+    if parsed_url.scheme != "file":
+        # we currently only support file:// style urls
+        raise PublishPathNotSupported(
+            "Cannot resolve unsupported url '%s' into a local path." % attachment_data["url"]
+        )
+
+    # file urls can be on the following standard form:
+    #
+    # Std unix path
+    # /path/to/some/file.txt -> file:///path/to/some/file.txt
+    #
+    # >>> urlparse.urlparse("file:///path/to/some/file.txt")
+    # ParseResult(scheme='file', netloc='', path='/path/to/some/file.txt', params='', query='', fragment='')
+    #
+    # windows UNC path
+    # \\laptop\My Documents\FileSchemeURIs.doc -> file://laptop/My%20Documents/FileSchemeURIs.doc
+    #
+    # >>> urlparse.urlparse("file://laptop/My%20Documents/FileSchemeURIs.doc")
+    # ParseResult(scheme='file', netloc='laptop', path='/My%20Documents/FileSchemeURIs.doc', params='', query='', fragment='')
+    #
+    # Windows path with drive letter
+    # C:\Documents and Settings\davris\FileSchemeURIs.doc -> file:///C:/Documents%20and%20Settings/davris/FileSchemeURIs.doc
+    #
+    # >>> urlparse.urlparse("file:///C:/Documents%20and%20Settings/davris/FileSchemeURIs.doc")
+    # ParseResult(scheme='file', netloc='', path='/C:/Documents%20and%20Settings/davris/FileSchemeURIs.doc', params='', query='', fragment='')
+    #
+    # for information about windows, see
+    # https://blogs.msdn.microsoft.com/ie/2006/12/06/file-uris-in-windows/
+
+    if parsed_url.netloc:
+        # unc path
+        resolved_path = urllib.unquote("//%s%s" % (parsed_url.netloc, parsed_url.path))
+    else:
+        resolved_path = urllib.unquote(parsed_url.path)
+
+    # python returns drive letter paths incorrectly and need adjusting.
+    if re.match("^/[A-Za-z]:/", resolved_path):
+        resolved_path = resolved_path[1:]
+
+    # we now have one of the following three forms (with slashes):
+    # /path/to/file.ext
+    # d:/path/to/file.ext
+    # //share/path/to/file.ext
+    log.debug("Path extracted from url: '%s'" % resolved_path)
+
+    # create a lookup table of shotgun paths,
+    # keyed by upper case storage name
+    log.debug("Building cross-platform path resolution lookup table:")
+    storage_lookup = {}
+    for storage in get_cached_local_storages(tk):
+        storage_key = storage["code"].upper()
+        storage_lookup[storage_key] = ShotgunPath.from_shotgun_dict(storage)
+        log.debug("Added Shotgun Storage %s: %s" % (storage_key, storage_lookup[storage_key]))
+
+    # get default environment variable set
+    # note that this may generate a None/None/None entry
+    storage_lookup["_DEFAULT_ENV_VAR_OVERRIDE"] = ShotgunPath(
+            os.environ.get("SHOTGUN_PATH_WINDOWS"),
```

Comment:
Would it be reasonable for someone to have multiple paths on a given platform, using os.pathsep between each? I know what I am saying is basically having multiple file storage.